examples/glium_gl_area/main.rs

```diff
@@ -0,0 +1,280 @@
+struct Vertex {
+    position: [f32; 2],
+    color: [f32; 3],
+}
+
+implement_vertex!(Vertex, position, color);
+
+struct Renderer {
+    context: GLAreaContext,
+    vertex_buffer: VertexBuffer<Vertex>,
+    index_buffer: IndexBuffer<u16>,
+    program: glium::Program,
+}
+
+impl Renderer {
+    /// # Safety
+    ///
+    /// Only call this from GLArea's realize and don't make any modifications to the GL context.
+    unsafe fn new(area: &gtk::GLArea) -> Option<Self> {
+        if area.error().is_some() {
+            return None;
+        }
+
+        let context = GLAreaContext::new(area.downgrade());
+
+        // The following code is from glium's triangle example:
+        // https://github.com/glium/glium/blob/2ff5a35f6b097889c154b42ad0233c6cdc6942f4/examples/triangle.rs
+        let vertex_buffer = VertexBuffer::new(
+            &context,
+            &[
+                Vertex {
+                    position: [-0.5, -0.5],
+                    color: [0., 1., 0.],
+                },
+                Vertex {
+                    position: [0., 0.5],
+                    color: [0., 0., 1.],
+                },
+                Vertex {
+                    position: [0.5, -0.5],
+                    color: [1., 0., 0.],
+                },
+            ],
+        )
+        .unwrap();
+        let index_buffer =
+            IndexBuffer::new(&context, PrimitiveType::TrianglesList, &[0u16, 1, 2]).unwrap();
+        let program = program!(&context,
+            140 => {
+                vertex: "
+                    #version 140
+                    uniform mat4 matrix;
+                    in vec2 position;
+                    in vec3 color;
+                    out vec3 vColor;
+                    void main() {
+                        gl_Position = vec4(position, 0.0, 1.0) * matrix;
+                        vColor = color;
+                    }
+                ",
+
+                fragment: "
+                    #version 140
+                    in vec3 vColor;
+                    out vec4 f_color;
+                    void main() {
+                        f_color = vec4(vColor, 1.0);
+                    }
+                "
+            },
+
+            110 => {
+                vertex: "
+                    #version 110
+                    uniform mat4 matrix;
+                    attribute vec2 position;
+                    attribute vec3 color;
+                    varying vec3 vColor;
+                    void main() {
+                        gl_Position = vec4(position, 0.0, 1.0) * matrix;
+                        vColor = color;
+                    }
+                ",
+
+                fragment: "
+                    #version 110
+                    varying vec3 vColor;
+                    void main() {
+                        gl_FragColor = vec4(vColor, 1.0);
+                    }
+                ",
+            },
+
+            100 => {
+                vertex: "
+                    #version 100
```

Comment:
What's the difference between the 3 shaders apart from the version? Maybe add a comment about that

Reply:
Took me a few reads to notice too. :) I'm not experienced enough in GLSL to describe the differences, I just took them verbatim from the glium example code.

What's the GTK minimum GL requirement? Maybe we can drop some of the older GLSL versions?

Reply:
If that's just taken from glium then let's keep it like that, I guess.

For GTK minimum version, no idea. @bilelmoussaoui ?

Reply:
No idea about this. Worth asking gtk developers about

Reply:
GTK is all over the place with its shaders, with many using version 150 and above. I think I'll leave just the 140 version in this example.

Reply:
@ebassi Maybe you know what the required minimum GLSL version for GTK4 is?

Reply:
The version of the shaders depend on the version of GL; currently, GTK4 uses:

 - 100 on GLES
 - 110 on GL2 legacy
 - 130 on GL3 legacy
 - 150 on GL3.2+

In practice, the version of GLSL for the shaders *inside your application* depends on the GL context you're either creating or using—i.e. if you support multiple versions of GL then you should load different shaders.

If you only care about recent GL, as you should, then going for GLSL 1.50 is perfectly fine; anything else will error out, and you can catch that error and fall back to something else.

Reply:
Thanks!

Reply:
Do you mind if I include this whole comment into that example?